### ISSUE TITLE
Hybrid lifestyle. What is wrong

### ISSUE THREAD

**Issue by vitidev** [simpleinjector/SimpleInjector#407]
I do not understand well why the combination of `Lifestyle.Scoped` and `Lifestyle.Transient` is caused a "lifestyle mismatch". After all, unlike a `Singleton`, scoped instanses are the same temporary as transient. 
And in general, the lifetime of an object depends on the retention references to instanse by someone. And therefore the aspect should not be on lifetime, but on the question "should a new object be created at the given place or re-used existing"

And although my opinion does not coincide with the opinion of the author, I do not want to turn off the option
Container.Options.SuppressLifestyleMismatchVerification

Downgrade `LifeTime.Scoped `-> `LifeTime.Transient` is meaningless, `Lifetime.Transient` -> `LifeTime.Scoped` causes errors. Thats why I use `Lifestyle.CreateHybrid`

``` c#
class ScopedService
{
    public ScopedService(IHybridDependency dep){}
}

public interface IHybridDependency{}
public class HybridDependency : IHybridDependency{}

static void Main(string[] args)
{
    var container=new Container();
    container.Options.DefaultScopedLifestyle=new AsyncScopedLifestyle();

    container.Register<ScopedService>(Lifestyle.Scoped);

    var hybridLifeStyle = Lifestyle.CreateHybrid(defaultLifestyle: Lifestyle.Scoped,
        fallbackLifestyle: Lifestyle.Transient);
    container.Register<IHybridDependency, HybridDependency>(hybridLifeStyle);

    using (AsyncScopedLifestyle.BeginScope(container))
    {
        var service = container.GetInstance<ScopedService>();
    }
    Console.ReadKey();
}
```

defaultLifestyle:  The lifestyle to use when its GetCurrentScope method returns a scope.. 
fallbackLifestyle:  The lifestyle to use when the GetCurrentScope method of the defaultLifestyle argument returns null. 

But I still get an error

> SimpleInjector.ActivationException: 'A lifestyle mismatch is encountered. ScopedService (Async Scoped) depends on IHybridDependency implemented by HybridDependency (Hybrid Scoped / Transient). Lifestyle mismatches can cause concurrency bugs in your application. Please see https://simpleinjector.org/dialm to understand this problem and how to solve it.'

Why the error?

Even more confusion from this

``` c#
container.Register(
  () => new ScopedService(container.GetInstance<IHybridDependency>()),
  Lifestyle.Scoped);
  
container.Register<IHybridDependency, HybridDependency>();

using (AsyncScopedLifestyle.BeginScope(container))
{
    var service = container.GetInstance<ScopedService>();
}
```

This registration simply bypasses verification

**Comment by dotnetjunkie** [comment]
> I do not understand well why the combination of `Lifestyle.Scoped` and `Lifestyle.Transient` is caused a "lifestyle mismatch".

Scopes can live for a very long time, even though in web applications a Scope is equal to the lifetime of a web request. Simple Injector however is framework agnostic and it can't assume that every scope you create runs in the context of a web request.

But even when ran within the context of a web request, injecting a `Transient` into a `Scoped` instance could lead to very subtle, very hard to find bugs, depending on how your application is designed. In Simple Injector we prefer being very explicit about this, and that's why by default this beahvior is disallowed.

What this means in general is that the `Transient` dependency actually is `Scoped`, and usually the fix is really simple: register the dependency as `Scoped` as well.

Do note however the difference between `Transient` and [Instance Per Dependency](https://simpleinjector.readthedocs.io/en/latest/lifetimes.html#instance-per-dependency). Tecnhically those lifestyles are the same, but with `Instance per Dependency` you state that you expect the dependency to live as long as its consumer. In other words, a `Instance Per Dependency` is a `Transient` dependency without checking for lifestyle mismatches.

Simple Injector out-of-the-box does not support `Instance per Dependency`, while Autofac has no support for `Transient`; it only supports `Instance per Dependency`.

Wanting to inject `Transient` instances into consumers with a longer lifestyle however, might indicate a design issue. This often indicates that the `Transient` dependency has _mutable state_, which might be the reason you don't want to share this state. In general however, your application components should be _immutable_. Runtime data should _flow_ through your application components using _method calls_, but state should not be stored into them.

Doing so can simplify the building of a system considerably, but is a different mind set, which is something not every developer is used to. But when an application component is immutable, it typically means that it doesn't have to be registered as `Transient`.

> But I still get an error

The reason you get this error is explained [here](https://simpleinjector.readthedocs.io/en/latest/LifestyleMismatches.html#what-about-hybrid-lifestyles):

> **As explained, components should only depend on equal length or longer lived components. But how long does a component with this hybrid lifestyle live? For components that are configured with the lifestyle defined above, it depends on the implementation of _someCondition_. But without taking this condition into consideration, we can say that it will at most live as long as the longest wrapped lifestyle [...] and at least live as long as shortest wrapped lifestyle [...].**

Since your hybrid lifestyle is a mix of both `Scoped` and `Transient`, Simple Injector will assume it to be `Transient` for diagnostics when it is used as a dependency, while assuming it to be `Scoped` when used as a consumer of some dependency. It has to do this, because it has no other way to determine that it would actually be safe to do so.

Long story short, you are trying to abbuse Hybrid lifestyles to suppress lifestyle mismatches. 

I'm glad to here you don't want to `SuppressLifestyleMismatchVerification` on a global level. This means that you see the benefits of this type of validation. 

My advise would be to make such component _immutable_ and register it as `Scoped`. If that isn't an option, you can also suppress Lifestyle Mismatches on an individual level. Although the [documentation](https://simpleinjector.readthedocs.io/en/latest/LifestyleMismatches.html#how-to-fix-violations) doesn't state this, you can absolutely do this. Reason it is not documented is because we feel that the need for this is trule exceptional. Again, having immutable components would typically solve this problem.

Suppressing warnings can be done in the same way as explained in [other parts of the documentation](https://simpleinjector.readthedocs.io/en/latest/disposabletransientcomponent.html#when-to-ignore-warnings), with the exception that you will have to suppress the warning on the _consumer_, since that's where the mismatch occurs. This means you will have to suppress the mismatch for _every consumer that uses this dependency_. 

The following code snippet shows how to suppress the warning:

``` c#
container.Register<ScopedService>(Lifestyle.Scoped); // depends on TransientDependency
container.Register<TransientDependency>(Lifestyle.Transient);

var registration = container.GetRegistration(typeof(ScopedService)).Registration;
registration.SuppressDiagnosticWarning(DiagnosticType.LifestyleMismatch);
```

**Comment by vitidev** [comment]
> Wanting to inject Transient instances into consumers with a longer lifestyle however, might indicate a design issue. This often indicates that the Transient dependency has mutable state, which might be the reason you don't want to share this state. 

Show me how to solve my typical problem without SuppressDiagnosticWarning in descktop application

we have:
1 TransientSrv - Short-live transient dependency aka *Repository/*Manager/(Provid|Read|Write)er/*/*/*
Does not contain a state. It is needed by other transient objects. Such objects are most and often defined without explicit registration
Thats why we register it as:

    container.Register<TransientSrv >(Lifestyle.Transient);

2 Several classes that are instantiated within scope during execution: SomeOperation1,  SomeOperation1,  SomeOperation3, etc. Scoped or transient - does not matter in this place

3 SharedDependency - shared for SomeOperation1,  SomeOperation2,  SomeOperation3 - must be scoped lifetime aka "singleton in scope". 

SharedDependency (scoped) has dependencies to TransientSrv (transient). I see no reason why it should not refer to TransientSrv

> and usually the fix is really simple: register the dependency as Scoped as well.

TransientSrv is also needed by others transient, so it can not have scoped lifetime

> Wanting to inject Transient instances into consumers with a longer lifestyle however, might indicate a design issue. 

Maybe I do not understand something, but in my opinion this is quite a typical situation

from documentation;

> Instead of injecting the dependency, inject a factory for the creation of that dependency and call that factory every time an instance is required.

If the factory have a scoped lifetime, it can't have transient dependency. If the factory have a transient lifetime - we get the same problem "lifestyle mismatch"

As a result, we can change the design or perform an explicit registration that bypasses the verification. But I do not see any problems in the design. It is very simple and typical

> Simple Injector out-of-the-box does not support Instance per Dependency

How to be in this case?

In theory, we can use CreateHybrid, but after your words

> Since your hybrid lifestyle is a mix of both Scoped and Transient, Simple Injector will assume it to be Transient for diagnostics when it is used as a dependency, while assuming it to be Scoped when used as a consumer of some dependency. It has to do this, because it has no other way to determine that it would actually be safe to do so.

I stopped understanding why he needed it and how to use it

**Comment by dotnetjunkie** [comment]
Most questions about Dependency Injection are very specific in nature. Since those questions are often coupled to the actual design and intend of the application, it makes it hard to answer those questions in a generic sense. I believe this holds in your case as well. Considering your very generic description of your application, I find it hard to imagine what's going on, what classes are doing, and how everything is wired together.

Because of this, do you mind providing me with a more detailed example of your application. Using real class names, its actual constructor, a short (one line) description of the intend of such class, and how things are wired up would be very useful.

> It is needed by other transient objects. Such objects are most and often defined without explicit registration

Is the fact that they are not registered explicitly the main reason for why those objects being transient, or do other requirements force them to be transient? Because if this is what you are facing, there are multiple ways to influence the lifestyle that those objects are gettings, such as:

 - Register them explicitly.
 - Set the `Container.Options.DefaultLifestyle` to a different lifestyle (is a global setting).
 - Replace the `Container.Options.LifestyleSelectionBehavior` with a custom implementation. This allows you to determine the lifestyle on a case-by-case basis.

> If the factory have a scoped lifetime, it can't have transient dependency

In Simple Injector, factories themselves can be singeltons without any problem. For instance:

``` c#
Func<MyTransient> factory = () => container.GetInstance<MyTransient>();

container.Register<MyTransient>(Lifestyle.Transient);
container.RegisterSingleton(factory);

var singletonFactory = container.GetInstance<Func<MyTransient>>();

var instance1 = singletonFactory.Invoke();
var instance2 = singletonFactory.Invoke();

Assert.AreNotSame(instance1, instance2);
```

**Comment by vitidev** [comment]
1 A very abstract example (without code)

Githab issue format does not involve large chunks of code, so I will draw a schematic

    SomeOperation1 \
    SomeOperation2-->MySharedScoped(scoped)-->MyTransient(transient) 
    SomeOperation3 /

You propose to use the factory(singleton), since it has a lifetime greater than scoped lifetime

    SomeOperation1 \
    SomeOperation2-->MySharedScoped(scoped)-->Factory(singleton)->MyTransient(transient) 
    SomeOperation3 /

This requires explicit registration

    Func<MyTransient> factory = () => container.GetInstance<MyTransient>();        
    container.RegisterSingleton(factory);                    
  
In doing so, we get some disadvantages
a) The long-lived object of the factory that is not needed outside the scopes
b) "MySharedScoped" now depends on the "Factory". In this case, the only reason for the existence of this factory is bypasses verification. And we need to use this factory in the constructor, that is bad
c) The factory should preferably be a class. Func<> factory as dependency is obscure 

Thus, the factory simply bypasses verification and creates a transient object once when creating "MySharedScoped"

But the same thing can be done with less code

    container.Register(()=>new MySharedScoped(container.Resolve<MyTransient>, ....));    
  
The disadvantage of this is only the need to change the registration when new dependency appear

The disadvantage of explicit registration is the bypass of verification
The documentation does not cover problems with scopes, only with singleton

2 More specific example (with pseudo-code. So shorter.)

    //class with  read write methods. TRANSIENT
    class DataReaderWriter {
        .ctor(string dataDir)      
    } 

    //statefull class. But needed as a singleton within only one large operation. So it has scoped lifetime
    class DataManager{
       .ctor(DataReaderWriter drw, ....)
    }

    //Many classes representing the operation itself. TRANSIENT
    class Preproccessor(){
       .ctor(DataManager dm, ....)
    }
    class Proccessor(){
       .ctor(DataManager dm, ....)
    }
    class PostProccessor(){
       .ctor(DataManager dm, ....)
    }
                   
Use within the same logical layer, where the container acts not as an antipattern service locator, but as a universal factory

    using (AsyncScopedLifestyle.BeginScope(container))
    {
        //start many nested methods  
        
        Somewhere inside inner methods
        preproccessor=container.GetInstance<Preproccessor>();  
        
        Somewhere inside another inner methods
        proccessor=container.GetInstance<Proccessor>()   
          
        Somewhere inside another inner methods
        postproccessor=container.GetInstance<PostProccessor>();
    }  

И understood the solution with a singleton factory for transient DataReaderWriter, but this solution has disadvantages and does not look exactly like compared to the explicit registration.

As for me container === universal factory. I use AsyncScopedLifestyle as an ambient factory context with scoped singletones and automatic dispose.

Another solution is to create a single AllFactory(scoped) for everything that is created inside the scope. Inside of AllFactory we decides who is the transient, and who is the singleton. But the creation of components in AllFactory will require much more manual code than the only explicit registration of a DataManager

AlFactory is a good choice in terms of clean design, but it practically does not use the container. But it is for the reduction of manual code and manual factories we choose IoC

3 You forgot to show me a real example of using the HybridLifetime. If he is not for solve these problems, then what is it for?

**Comment by dotnetjunkie** [comment]
It seems that the `DataManager` takes a hard dependency on the `DataReaderWriter` and that it depends on its mutability and the state changes that it does, and it seems that `DataManager` can't function correctly when `DataReaderWriter` is shared across multiple instances.

This is something I find suspicious. Questions: Why does the `DataManager` have:

 1. a dependency on a concrete type `DataReaderWriter`, and
 2. isn't itself in control over the lifetime of this `DataReaderWriter`?

`DataReaderWriter` seems like a candidate for being an implementation detail to the `DataManager`. Why don't you let the `DataManager` create it internally?


> You forgot to show me a real example of using the HybridLifetime. If he is not for solve these problems, then what is it for?

Examples of such can be found in the [documentation](https://simpleinjector.readthedocs.io/en/latest/lifetimes.html#hybrid). Hybrid lifestyles allow switching to different kinds of contexts, for instance when you have components run within a web request (using `WebRequestLifestyle` for instance) while you need to run those same components on a background thread, where there is no Web Request. Since we are moving more and more to one single type of scope (primarily `AsyncScopedLifestyle`), the we will need to use Hybrid lifestyles less and less, since `AsyncScopedLifestyle` can be started explicitly, which means that we can run both web requests and background operations using the same scoped lifestyle.

**Comment by vitidev** [comment]
While I was writing a lot of explanations ... I understood "what's wrong"

"LifeTime mismatch" means "**potentially** dangerous in multithreaded mode"

But is there any danger if
1) The code is guaranteed to be executed in single-threaded mode
2) transient dependency is stateless without thread-unsafe dependies within
3) transient dependency is thread-safe

In addition, +2 rules of my programming style (Does anyone program differently?):

4) The code knows that it is running in multithreaded mode, and it itself protects dependencies
5) Or make sure that the dependencies are thread-safe
(The rule has some danger that thread-safe dependencies will become dangerous, but in my case this will be considered a design error)

There is no design error. "LifeTime mismatch" is not an error, but the signal "attention to dependencies"
If you follow the rules, then there is no need for crutches like the singleton factory.


How to deal with error "lifetime mismatch" (without exhausting explicit registration)?

1) Disable verification (Container.Options.SuppressLifestyleMismatchVerification) completely (too radical)

2) Disable check for lifetime (still too radical)  
                   
      var registration = container.GetRegistration(typeof(ScopedService)).Registration;
      registration.SuppressDiagnosticWarning(DiagnosticType.LifestyleMismatch);           
    
3) The ability to specify "yes I'm sure that this transient dependence can safely be included in other lifetime" (good solution)

     May be add a new type of lifetime "ThreadSafeTransient"?

4) The ability to specify "I wrote this code and I'm sure that this class itself protects all the transient dependencies" (good solution)

May be add a new type of lifetime "ThreadSafeSingleton(Scope)"?


How to implement 3 and 4?

--------------------------

3) The ability to specify "yes I'm sure that this transient dependence can safely be included in other lifetime" (good solution)

according to you

> Hybrid lifestyles allow switching to different kinds of contexts, for instance when you have components run within a web request (using WebRequestLifestyle for instance) while you need to run those same components on a background thread, where there is no Web Request.

        var hybridLifeStyle = Lifestyle.CreateHybrid(defaultLifestyle: Lifestyle.Scoped,
        fallbackLifestyle: Lifestyle.Transient);
                                                      
But Simple Injector prevents me from doing this without verification disabled (And when the verification is disabled -the hybrid lifestyle is not needed )      

> Since your hybrid lifestyle is a mix of both Scoped and Transient, Simple Injector will assume it to be Transient for diagnostics when it is used as a dependency,
---------------------
>  It has to do this, because it has no other way to determine that it would actually be safe to do so.

Perhaps the introduction of new lifetime will be "other way to determine that it would actually be safe"

That is, it will be a kind of flag "I received a signal and made sure everything is safe"
Also, do not forget the  the support in custom lifetime

**Comment by dotnetjunkie** [comment]
> But is there any danger if
> - The code is guaranteed to be executed in single-threaded mode

If the dependency is mutable, it depends on what consumers expect whether Transient would actually work out well. Good example is a `DbContext`. This should typically not be Transient at all, but Scoped.

> - transient dependency is stateless without thread-unsafe depended within
> - transient dependency is thread-safe

That would be safe, but in these two cases the dependency can be even promoted to either Scoped or Singleton. That's why the Simple Injector documentation promotes this as solution.

> The code knows that it is running in multithreaded mode, and it itself protects dependencies

I certainly program differently. My application codes do not do any multi-threaded code. Any knowledge about the application running multi-threaded is centralized in the Composition Root. Take a look at this [documentation page](https://simpleinjector.readthedocs.io/en/latest/howto.html#work-with-dependency-injection-in-multi-threaded-applications) that describes this.

> There is no design error. "LifeTime mismatch" is not an error,

True, but it is a very strong signal, and is typically something that can be fixed.

> How to deal with error "lifetime mismatch" (without exhausting explicit registration)?

But why do you have that many implicit concrete registrations anyway? And if you do, why not override the lifestyle behavior on the container for these implicit registrations?

> But Simple Injector prevents me from doing this without verification disabled

Not at all. A Transient/Scoped hybrid lifestyle can safely be injected into a transient consumer, without getting a warning from Simple Injector. Again, you can't use hybrids to suppress these warnings, and as I said, that's not what they are created for.

> May be add a new type of lifetime "ThreadSafeTransient"?
> The ability to specify "I wrote this code and I'm sure that this class itself protects all the transient dependencies" (good solution)
> The ability to specify "yes I'm sure that this transient dependence can safely be included in other lifetime" (good solution)

You mean a customly built [InstancePerDependencyLifestyle](https://github.com/simpleinjector/SimpleInjector/blob/master/src/SimpleInjector.CodeSamples/InstancePerDependencyLifestyle.cs).

> Also, do not forget the the support in custom lifetime

The lack of built-in Instance Per Dependency Lifestyle is deliberate. This doesn't fit the Simple Injector philosophy, which is to [steer developers into (what we feel is) best practices](https://simpleinjector.readthedocs.io/en/latest/principles.html#push-developers-into-best-practices). We like to be able to [fail fast](https://simpleinjector.readthedocs.io/en/latest/principles.html#never-fail-silently). We rather provide developers with good guidance on how to change their design accordingly (although this might involve forum discussions) and instead of creating 'dangerous' things out of the box, provide the proper extension points and [examples](https://github.com/simpleinjector/SimpleInjector/blob/master/src/SimpleInjector.CodeSamples/InstancePerDependencyLifestyle.cs) that show you how to shoot yourself in the foot :).

**Comment by vitidev** [comment]
> You mean a customly built InstancePerDependencyLifestyle.

Looks like this is what I need

> Not at all. A Transient/Scoped hybrid lifestyle can safely be injected into a transient consumer, without getting a warning from Simple Injector. 

But cannot be injected into a scoped consumer in my case. 

I think I understood my mistake.
The result lifetime for a hybrid lifetime depends from container.Options.DefaultScopedLifestyle, but not from local context during call GetInstanse<>? 

> But why do you have that many implicit concrete registrations anyway? And if you do, why not override the lifestyle behavior on the container for these implicit registrations?

I'm not sure I understand you.
1) SI does not require explicit registration as autofac. it is very convenient. Also, you do not need to change the registration when changing the signature of the constructor.
2) Explicit registration bypasses lifetime diagnostic

         container.Register(
              ()=>new MyService( //Scoped
                    container.GetInstanse<> //Transient
               ) //Scoped
         , LifeTime.Scoped)


--------------------
On the remaining arguments I will answer when I thoroughly examine your arguments and links. I do not agree with you in everything, but I do not have enough understanding for an exact argumentation.  Also poor knowledge of the language prevents me from expressing the idea accurately.

**Comment by dotnetjunkie** [comment]
> SI does not require explicit registration as autofac. it is very convenient. Also, you do not need to change the registration when changing the signature of the constructor.

Yes, but this is something that is probably change in v5. Although it can be very convenient to have some concrete unregistered types, this is the most common source of errors and confusion. See [this discussion](https://github.com/simpleinjector/SimpleInjector/issues/377) for more information.

> The result lifetime for a hybrid lifetime depends from container.Options.DefaultScopedLifestyle

Well, not completely. A hybrid lifestyle doesn't have to depend on  `Options.DefaultScopedLifestyle`. For instance, you can create a Singleton/Transient hybrid. But even if you do create a WebRequest/AsyncScoped hybrid, you don't need to set it as `DefaultScopedLifestyle` (although this would be the most common thing to do). Typically however, a hybrid lifestyle contains at least one scoped lifestyle and Simple Injector will check for `ScopedLifestyle.GerCurrentScope() != null`. It that expression returns false, the `fallbackLifestyle` will be used. So again, no notion of `DefaultScopedLifestyle` here.


> Explicit registration bypasses lifetime diagnostic

Yes, such construct completely blinds the diagnostic services and disallows certain optimizations. It's not adviced because of that.

> poor knowledge of the language prevents me from expressing the idea accurately.

Don't worry. This stuff isn't easy, even if you were a native speaker, expressing this stuff through an online forum isn't that simple at all.


One thing, in your case do change this line:

``` c#
public override int Length => Singleton.Length;
```

to:

``` c#
public override int Length => Scoped.Length;
```

This allows the `InstancePerDependency` registrations to be injected into `Scoped` registrations, but not into `Singletons`.

**Comment by vitidev** [comment]
Correct me if I am wrong

I suspect that ScopedLifestyle.GerCurrentScope() != null use within CallContext.LogicalGetData and my code from first message of the issue must be work..but is not work, becouse...

If we need to use a transient lifetime inside the scope lifetime, we can use the hybrid as a **temporary** solution.

However,  SI considers the hybrid lifetime as a bad solution and causes "lifetime mismatch" **to force** find another solution.

To use the temporary solution, we are forced to turn off the validation and upon activation, the error will again be called as a **reminder** of the unsolved problem.

**Comment by vitidev** [comment]
We discussed too many subjects and I'm confused. Therefore, I will ask an important question.

What problems can be from depending scoped-lifetime from transient in **single**-threaded application?

**Comment by dotnetjunkie** [comment]
Can you rephrase your question? I'm not sure what you are asking.

**Comment by vitidev** [comment]
**Transient**

Transient is a short-lived object that is created every time it is accessed through si. For good design, a transient object does not need to know about a multithreaded context 
And since it registers in a global container, then it can be requested from anywhere.

Child containers do not support by si.

**Singleton <-> Transient**

Based on the article http://blog.ploeh.dk/2014/06/02/captive-dependency/, we define the problem:

If a transient is captured by a singleton, then transient become an implicit singleton. Singletons are the same **superglobal ever**-living objects, for which the probability of simultaneous access to the captured transeint from different thread is great.

Of course, transient object can be read-only/immutable and be thread-safe, it can not know about this and the probability of error is high

**Scoped <-> Transient**

Is an invariant of PerGraph just stretched over time.

1) ScopedLifetime is "singelton in scope". ScopedLifetime certainly has a longer lifetime than "transient" and **can have** the same problems as the global singleton.

2) ScopedLifetime **does not exist** outside of the scope. Therefore, it is impossible to make ScopedLifetime as Singleton, and it is also useless to make a singleton factory - becouse scoped is is singelton(in scope)

3) https://simpleinjector.readthedocs.io/en/latest/lifetimes.html#scoped All five(Where is 5???) lifestyles are performed in a certain context (scope). 

    And my scope usage

        Parallel.Foreach(
           using(scope begin){   
           	await   	
           	await
           }
        )

  That is, although execution flow can jump from a thread to a thread, but there is no simultaneous multithreaded access and race condition. And the problem described in the section "Singleton <-> Transient" will appear only if they inside will use the threads themselves.


**conclusions:**

1) If there are no **other** side effects, then this is not a bad design.

2) I can not use the InstancePerDependencyLifestyle, because only the Scoped registration knows that the injection of the other transient is safe. And I do not have the ability to specify this, so I have to use "bypass verdication". 	
It's a pity!
	
3) I see no way to "fix" the "problem", because your proposal "make a singleton or singleton factory" is impossible.

**Comment by vitidev** [comment]
As a result, we have that ideology completely does not accept the concept "local singleton" not suggesting any solution to the problem.

**Comment by vitidev** [comment]
It is difficult to formulate ideas because of the language barrier
Therefore, I quote from the autofac wiki

>   Given the developer of the application is ultimately responsible for determining whether captives are OK or not, the developer may determine that it’s acceptable for a singleton, for example, to take an “instance-per-dependency” service.
For example, maybe you have a caching class that is intentionally set up to cache things for only the lifetime of the consuming component. If the consumer is a singleton, the cache can be used to store things for the whole app lifetime; if the consumer is “instance-per-request” then it only stores data for a single web request. In a case like that, you may end up with a longer-lived component taking a dependency on a shorter-lived component intentionally.
This is acceptable as long as the application developer understands the consequences of setting things up with such lifetimes. Which is to say, if you’re going to do it, do it intentionally rather than accidentally.

So, if I have a thread-safe caching " caching class", then I can not register it as a singleton. Also it cannot be registered as transient without switch off Container.Options.SuppressLifestyleMismatchVerification, because simple injector does not have built-in tools to solve this problem, like "NonCaptive/SafeTransient" lifestyle

I would be glad to see the solution of the problem " non-captive caching class" with simple injector.

**Comment by dotnetjunkie** [comment]
You can supress the warning by calling `SuppressDiagnosticWarning` on the registration, similar to shown here: https://simpleinjector.readthedocs.io/en/latest/disposabletransientcomponent.html
